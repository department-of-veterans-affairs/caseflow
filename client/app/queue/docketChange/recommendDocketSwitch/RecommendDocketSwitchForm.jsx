import React from 'react';
import PropTypes from 'prop-types';

import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers';
import * as yup from 'yup';

import { DOCKET_SWITCH_RECOMMENDATION_TITLE } from '../../../../COPY';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import Button from '../../../components/Button';
import { sprintf } from 'sprintf-js';
import { dispositionOptions, dispositions } from '../constants';
import TextareaField from '../../../components/TextareaField';
import RadioField from '../../../components/RadioField';

const schema = yup.object().shape({
  summary: yup.string(),
  timely: yup.string().required(),
  disposition: yup.
    mixed().
    oneOf(dispositions).
    required(),
  hyperlink: yup.string(),
});

const TIMELY_OPTIONS = [
  { displayText: 'Yes', value: 'yes' },
  { displayText: 'No', value: 'no' },
];

export const RecommendDocketSwitchForm = ({
  claimantName,
  onCancel,
  onSubmit,
}) => {
  const { register, handleSubmit, errors, formState } = useForm({
    resolver: yupResolver(schema),
    mode: 'onBlur',
  });

  return (
    <form
      className="docket-change-recommendation"
      onSubmit={handleSubmit(onSubmit)}
    >
      <AppSegment filledBackground>
        <h1>{sprintf(DOCKET_SWITCH_RECOMMENDATION_TITLE, claimantName)}</h1>

        <TextareaField
          inputRef={register}
          name="summary"
          label="Add a summary of the request to switch dockets:"
          strongLabel
        />

        <RadioField
          name="timely"
          label="Is this request timely?"
          options={TIMELY_OPTIONS}
          inputRef={register}
          strongLabel
          vertical
        />

        <RadioField
          name="disposition"
          label="What is your recommendattion for this request to switch dockets?"
          options={dispositionOptions}
          inputRef={register}
          strongLabel
          vertical
        />
      </AppSegment>
      <div className="controls cf-app-segment">
        <Button
          type="submit"
          name="submit"
          disabled={!formState.isValid}
          classNames={['cf-right-side']}
        >
          Submit
        </Button>
        {onCancel && (
          <Button
            type="button"
            name="Cancel"
            classNames={['cf-right-side', 'usa-button-secondary']}
            onClick={onCancel}
            styling={{ style: { marginRight: '1rem' } }}
          >
            Cancel
          </Button>
        )}
      </div>
    </form>
  );
};

RecommendDocketSwitchForm.propTypes = {
  claimantName: PropTypes.string.isRequired,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
};
