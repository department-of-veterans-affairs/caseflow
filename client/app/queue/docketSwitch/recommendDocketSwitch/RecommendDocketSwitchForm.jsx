import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';

import { useForm, Controller } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';

import {
  DOCKET_SWITCH_RECOMMENDATION_TITLE,
  DOCKET_SWITCH_RECOMMENDATION_INSTRUCTIONS,
} from '../../../../COPY';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import Button from '../../../components/Button';
import { sprintf } from 'sprintf-js';
import DISPOSITIONS from '../../../../constants/DOCKET_SWITCH_DISPOSITIONS';
import TextField from '../../../components/TextField';
import TextareaField from '../../../components/TextareaField';
import RadioField from '../../../components/RadioField';
import SearchableDropdown from '../../../components/SearchableDropdown';

const schema = yup.object().shape({
  summary: yup.string().required(),
  timely: yup.string().required(),
  disposition: yup.
    mixed().
    oneOf(Object.keys(DISPOSITIONS)).
    required(),
  hyperlink: yup.string(),
  judge: yup.
    object().
    shape({ label: yup.string().required(), value: yup.number().required() }).
    required(),
});

const TIMELY_OPTIONS = [
  { displayText: 'Yes', value: 'yes' },
  { displayText: 'No', value: 'no' },
];

export const RecommendDocketSwitchForm = ({
  appellantName,
  defaultJudgeId,
  judgeOptions = [],
  onCancel,
  onSubmit,
}) => {
  const { register, handleSubmit, formState, control, setValue } = useForm({
    resolver: yupResolver(schema),
    mode: 'onChange',
  });

  // The `SearchableDropdown` component uses objects as values, so here we determine that if defaultJudgeId is set
  const judgeDefault = useMemo(
    () =>
      judgeOptions.find(
        (opt) => defaultJudgeId && opt.value === defaultJudgeId
      ),
    [defaultJudgeId, judgeOptions]
  );
  const dispositionOptions = useMemo(() => Object.values(DISPOSITIONS), []);

  useEffect(() => {
    setValue('judge', judgeDefault);
  }, [judgeDefault]);

  return (
    <form
      className="docket-switch-recommendation"
      onSubmit={handleSubmit(onSubmit)}
      aria-label="Recommendation on Docket Switch"
    >
      <AppSegment filledBackground>
        <h1>{sprintf(DOCKET_SWITCH_RECOMMENDATION_TITLE, appellantName)}</h1>
        <p style={{ marginBottom: '30px' }}>
          {DOCKET_SWITCH_RECOMMENDATION_INSTRUCTIONS}
        </p>

        <TextareaField
          inputRef={register}
          name="summary"
          label="Add a summary of the request to switch dockets:"
          strongLabel
        />

        <RadioField
          name="timely"
          label="Is this request timely?"
          options={TIMELY_OPTIONS}
          inputRef={register}
          strongLabel
          vertical
        />

        <RadioField
          name="disposition"
          label="What is your recommendation for this request to switch dockets?"
          options={dispositionOptions}
          inputRef={register}
          strongLabel
          vertical
        />

        <TextField
          inputRef={register}
          name="hyperlink"
          label="Insert hyperlink to draft letter"
          strongLabel
        />

        <div style={{ maxWidth: '35rem' }}>
          <Controller
            as={SearchableDropdown}
            control={control}
            defaultValue={judgeDefault}
            name="judge"
            label="Assign to judge"
            searchable
            options={judgeOptions}
            placeholder="Select judge"
            strongLabel
          />
        </div>
      </AppSegment>
      <div className="controls cf-app-segment">
        <Button
          type="submit"
          name="submit"
          disabled={!formState.isValid}
          classNames={['cf-right-side']}
        >
          Submit
        </Button>
        {onCancel && (
          <Button
            type="button"
            name="Cancel"
            classNames={['cf-right-side', 'usa-button-secondary']}
            onClick={onCancel}
            styling={{ style: { marginRight: '1rem' } }}
          >
            Cancel
          </Button>
        )}
      </div>
    </form>
  );
};

RecommendDocketSwitchForm.propTypes = {
  appellantName: PropTypes.string.isRequired,
  defaultJudgeId: PropTypes.number,
  judgeOptions: PropTypes.array.isRequired,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
};
