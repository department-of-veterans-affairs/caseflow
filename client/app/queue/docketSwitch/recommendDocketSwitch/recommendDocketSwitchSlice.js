import { createAsyncThunk } from '@reduxjs/toolkit';
import ApiUtil from '../../../util/ApiUtil';

export const createDocketSwitchRulingTask = createAsyncThunk(
  'tasks/createDocketSwitchRulingTask',
  async (data) => {
    try {
      const res = await ApiUtil.post('/tasks', { data });
      const updatedTasks = res.body?.tasks?.data;

      return updatedTasks;
    } catch (error) {
      console.error('Error creating task', error);
      throw error;
    }
  }
);
