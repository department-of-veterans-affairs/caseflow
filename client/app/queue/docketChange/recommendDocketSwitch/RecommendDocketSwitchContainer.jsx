import React, { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router';
import { fetchJudges } from '../../QueueActions';

import { taskById, appealWithDetailSelector } from '../../selectors';
import { RecommendDocketSwitchForm } from './RecommendDocketSwitchForm';

export const RecommendDocketSwitchContainer = () => {
  const { taskId, appealId } = useParams();
  const { goBack } = useHistory();
  const dispatch = useDispatch();

  const task = useSelector((state) => taskById(state, { taskId }));
  const appeal = useSelector((state) =>
    appealWithDetailSelector(state, { appealId })
  );
  const judges = useSelector((state) => state.queue.judges);
  const judgeOptions = useMemo(
    () =>
      Object.values(judges).map(({ id: value, display_name: label }) => ({
        label,
        value,
      })),
    [judges]
  );

  // TODO - add logic to pull this from task tree
  const defaultJudgeId = useMemo(() => 3, [judges, appeal]);

  const handleSubmit = (formData) => console.log('handleSubmit', formData);

  useEffect(() => {
    if (!judgeOptions.length) {
      dispatch(fetchJudges());
    }
  });

  return (
    <RecommendDocketSwitchForm
      onCancel={goBack}
      onSubmit={handleSubmit}
      judgeOptions={judgeOptions}
      defaultJudgeId={defaultJudgeId}
    />
  );
};
