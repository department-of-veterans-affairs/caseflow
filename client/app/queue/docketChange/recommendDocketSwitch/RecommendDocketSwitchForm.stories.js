import React from 'react';

import { RecommendDocketSwitchForm } from './RecommendDocketSwitchForm';

export default {
  title: 'Queue/Docket Change/RecommendDocketSwitchForm',
  component: RecommendDocketSwitchForm,
  decorators: [],
  parameters: {},
  args: {
    claimantName: 'Jane Doe',
  },
  argTypes: {
    onCancel: { action: 'cancel' },
    onSubmit: { action: 'submit' },
  },
};

const Template = (args) => <RecommendDocketSwitchForm {...args} />;

export const Basic = Template.bind({});

Basic.parameters = {
  docs: {
    storyDescription:
      'Used by attorney at Clerk of the Board office to recommend a ruling to VLJ/Clerk',
  },
};
