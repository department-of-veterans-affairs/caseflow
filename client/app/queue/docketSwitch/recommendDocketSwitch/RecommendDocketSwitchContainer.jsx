import React, { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router';
import { fetchJudges } from 'app/queue/QueueActions';

import { appealWithDetailSelector, distributionTasksForAppeal, rootTasksForAppeal } from '../../selectors';
import DISPOSITIONS from 'constants/DOCKET_SWITCH_DISPOSITIONS';

import { RecommendDocketSwitchForm } from './RecommendDocketSwitchForm';
import {
  DOCKET_SWITCH_RECOMMENDATION_SUCCESS_TITLE,
  DOCKET_SWITCH_RECOMMENDATION_SUCCESS_MESSAGE,
} from 'app/../COPY';

import { sprintf } from 'sprintf-js';
import { showSuccessMessage } from 'app/queue/uiReducer/uiActions';
import { completeTask, createDocketSwitchRulingTask } from '../docketSwitchSlice';

// This takes form data and generates Markdown-formatted text to be saved as task instructions
export const formatDocketSwitchRecommendation = ({
  summary,
  timely,
  disposition,
  hyperlink,
}) => {
  const parts = [];

  const timelyCaps = timely[0].toUpperCase() + timely.substring(1);

  parts.push(`**Summary:** ${summary}`);
  parts.push(`**Is this a timely request:** ${timelyCaps}`);
  parts.push(`**Recommendation:** ${DISPOSITIONS[disposition].displayText}`);
  if (hyperlink) {
    parts.push(`**Draft letter:** [View link](${hyperlink})`);
  }

  // Separate each chunk by two line breaks
  return parts.join('  \n  \n');
};

export const RecommendDocketSwitchContainer = () => {
  const { appealId, taskId } = useParams();
  const { goBack, push } = useHistory();
  const dispatch = useDispatch();

  const appeal = useSelector((state) => appealWithDetailSelector(state, { appealId }));
  const rootTask = useSelector((state) => rootTasksForAppeal(state, { appealId }))[0];
  const distributionTask = useSelector((state) => distributionTasksForAppeal(state, { appealId }))[0];

  const judges = useSelector((state) => state.queue.judges);
  const judgeOptions = useMemo(
    () =>
      Object.values(judges).map(({ id: value, display_name: label }) => ({
        label,
        value,
      })),
    [judges]
  );

  // We want to default the judge selection to the VLJ currently assigned to the case, if exists
  const defaultJudgeId = useMemo(() => {
    // eslint-disable-next-line no-undefined
    return appeal.assignedJudge?.id ?? undefined;
  }, [judges, appeal]);

  // eslint-disable-next-line no-console
  const handleSubmit = async (formData) => {

    const instructions = formatDocketSwitchRecommendation({ ...formData });
    const newTask = {
      parent_id: appeal.distributedToJudge ? rootTask.taskId : distributionTask.taskId,
      type: 'DocketSwitchRulingTask',
      external_id: appeal.externalId,
      instructions,
      assigned_to_id: formData.judge.value,
      assigned_to_type: 'User',
    };

    const data = {
      tasks: [newTask],
    };

    const successMessage = {
      title: sprintf(DOCKET_SWITCH_RECOMMENDATION_SUCCESS_TITLE, appeal.appellantFullName, formData.judge.label),
      detail: DOCKET_SWITCH_RECOMMENDATION_SUCCESS_MESSAGE,
    };

    try {
      await dispatch(createDocketSwitchRulingTask(data));

      await dispatch(completeTask({ taskId }));

      dispatch(showSuccessMessage(successMessage));
      push('/queue');
    } catch (error) {
      // Perhaps show an alert that indicates error, advise trying again...?
      console.error('Error saving task', error);
    }
  };

  useEffect(() => {
    if (!judgeOptions.length) {
      dispatch(fetchJudges());
    }
  });

  return (
    <RecommendDocketSwitchForm
      onCancel={goBack}
      onSubmit={handleSubmit}
      judgeOptions={judgeOptions}
      defaultJudgeId={defaultJudgeId}
      appellantName={appeal.appellantFullName}
    />
  );
};
