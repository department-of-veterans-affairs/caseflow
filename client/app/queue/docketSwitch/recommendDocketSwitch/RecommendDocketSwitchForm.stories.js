import React from 'react';

import { RecommendDocketSwitchForm } from './RecommendDocketSwitchForm';

const judgeOptions = [
  { value: 1, label: 'Judge 1' },
  { value: 2, label: 'Judge 2' },
];

export default {
  title: 'Queue/Docket Switch/RecommendDocketSwitchForm',
  component: RecommendDocketSwitchForm,
  decorators: [],
  parameters: {},
  args: {
    appellantName: 'Jane Doe',
    judgeOptions,
  },
  argTypes: {
    onCancel: { action: 'cancel' },
    onSubmit: { action: 'submit' },
  },
};

const Template = (args) => <RecommendDocketSwitchForm {...args} />;

export const Basic = Template.bind({});

Basic.parameters = {
  docs: {
    storyDescription:
      'Used by attorney at Clerk of the Board office to recommend a ruling to VLJ/Clerk',
  },
};

export const DefaultJudge = Template.bind({});
DefaultJudge.args = {
  defaultJudgeId: 2
};
