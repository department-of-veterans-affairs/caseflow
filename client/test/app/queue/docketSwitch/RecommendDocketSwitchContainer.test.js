import { formatDocketSwitchRecommendation } from 'app/queue/docketSwitch/recommendDocketSwitch/RecommendDocketSwitchContainer';

describe('formatDocketSwitchRecommendation', () => {
  const summary = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit...';
  let timely = 'yes';
  const hyperlink = 'https://example.com/file.txt';

  let disposition = 'granted';

  describe('with granted disposition', () => {
    beforeAll(() => {
      disposition = 'granted';
    });

    it('properly formats', () => {
      const res = formatDocketSwitchRecommendation({
        summary,
        timely,
        hyperlink,
        disposition,
      });

      expect(res).toMatch(
        new RegExp(`\\*\\*Summary:\\*\\* ${summary} {2}\\n {2}\\n`)
      );
      expect(res).toMatch(
        new RegExp('\\*\\*Is this a timely request:\\*\\* Yes {2}\\n {2}\\n')
      );
      expect(res).toMatch(
        new RegExp('\\*\\*Recommendation:\\*\\* Grant all issues {2}\\n {2}\\n')
      );
      expect(res).toMatch(`**Draft letter:** [View link](${hyperlink})`);
      expect(res).toMatchSnapshot();
    });
  });

  describe('with partially_granted disposition', () => {
    beforeAll(() => {
      disposition = 'partially_granted';
    });

    it('properly formats', () => {
      const res = formatDocketSwitchRecommendation({
        summary,
        timely,
        hyperlink,
        disposition,
      });

      expect(res).toMatch(
        new RegExp(`\\*\\*Summary:\\*\\* ${summary} {2}\\n {2}\\n`)
      );
      expect(res).toMatch(
        new RegExp('\\*\\*Is this a timely request:\\*\\* Yes {2}\\n {2}\\n')
      );
      expect(res).toMatch(
        new RegExp(
          '\\*\\*Recommendation:\\*\\* Grant a partial switch {2}\\n {2}\\n'
        )
      );
      expect(res).toMatch(`**Draft letter:** [View link](${hyperlink})`);
      expect(res).toMatchSnapshot();
    });
  });

  describe('with denied disposition', () => {
    beforeAll(() => {
      disposition = 'denied';
    });

    it('properly formats', () => {
      const res = formatDocketSwitchRecommendation({
        summary,
        timely,
        hyperlink,
        disposition,
      });

      expect(res).toMatch(
        new RegExp(`\\*\\*Summary:\\*\\* ${summary} {2}\\n {2}\\n`)
      );
      expect(res).toMatch(
        new RegExp('\\*\\*Is this a timely request:\\*\\* Yes {2}\\n {2}\\n')
      );
      expect(res).toMatch(
        new RegExp('\\*\\*Recommendation:\\*\\* Deny all issues {2}\\n {2}\\n')
      );
      expect(res).toMatch(`**Draft letter:** [View link](${hyperlink})`);
      expect(res).toMatchSnapshot();
    });
  });
});

