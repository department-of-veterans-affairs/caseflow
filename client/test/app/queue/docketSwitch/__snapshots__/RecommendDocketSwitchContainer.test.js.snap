// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDocketSwitchRecommendation with denied disposition properly formats 1`] = `
"**Summary:** Lorem ipsum dolor sit amet, consectetur adipiscing elit...  
  
**Is this a timely request:** Yes  
  
**Recommendation:** Deny all issues  
  
**Draft letter:** [View link](https://example.com/file.txt)"
`;

exports[`formatDocketSwitchRecommendation with granted disposition properly formats 1`] = `
"**Summary:** Lorem ipsum dolor sit amet, consectetur adipiscing elit...  
  
**Is this a timely request:** Yes  
  
**Recommendation:** Grant all issues  
  
**Draft letter:** [View link](https://example.com/file.txt)"
`;

exports[`formatDocketSwitchRecommendation with partially_granted disposition properly formats 1`] = `
"**Summary:** Lorem ipsum dolor sit amet, consectetur adipiscing elit...  
  
**Is this a timely request:** Yes  
  
**Recommendation:** Grant a partial switch  
  
**Draft letter:** [View link](https://example.com/file.txt)"
`;
